### DISCUSSION TITLE
PDF in Windows Air application

### DISCUSSION THREAD

**Discussion by PetVermeulen** [airsdk/Adobe-Runtime-Support#1411] [Q&A]
Hi,

we have an Air based Windows application that is used by several of our customers. Inside this application, we show PDF documents by loading them into a mx:HTML element. This worked fine, until now...
With the upgrade by Adobe of its Acrobat Reader to 64-bit, this no longer works. As soon as we try to load a PDF document inside the HTML element, the complete application crashes.

Does anyone has an idea how to solve this issue?

Thanks a lot for your help!

regards,
Peter, Belgium

**Comment by jdomingo-ob** [comment]
Same problem, we could not load pdf via HtmlLoader, with older pdf reader (32bits) works, but some clients using other readers reported the same problem, and mac users hasn't the 32 bits option, any workarround?

**Reply by ajwfrost** [reply]
I've reached out to Adobe to see if I can get a contact within the Acrobat team to ask about this... there are some "x86" libraries that are deployed along with the 64-bit version of Acrobat Reader but none of them seem to provide the plug-in interface.

FWIW we may be switching to 64-bit AIR as the default version, and that might solve things (not sure whether you can try already testing it by generating your app with `-arch x64` when you build for Windows?)

thanks

**Comment by concord232** [comment]
Facing the same issue, @jdomingo-ob  @PetVermeulen did you guys find any solution?

I tried running application as 64 bit but that didn't help

**Reply by ranzou06** [reply]
At this point the only solution is to install the 32bit version of Acrobat, I doubt we well ever have a fix for that on the reader.

**Comment by MalacTheLittle** [comment]
Maybe it's worth try adding architecture tag in application descriptor...
`<architecture>64</architecture>`

**Reply by concord232** [reply]
I did try, 

I added <architecture>64</architecture> in the descriptor file and added "-arch x64" to adt while creating the package.

the created exe (when using v32 SDK) crashes when trying to render PDF, event the debugger also crashes.
created exe (when using v33.1 SDK) fails to even launch.

**Comment by pcollardez** [comment]
Still not working with properly compiled 64 bits app (SDK 33.1.1.713, windows, captive, Acrobat Reader/Pro 22.002.20191)

Remark: newer SDK breaks mx:HTML entirely.

According to the debugger, the crash occurs in the following library:

C:\Program Files\Adobe\Acrobat DC\Acrobat\Air\nppdf32.dll

Even if it is named "32", it seems to be a 64 bits module.

Message trapped by the debugger:

> Unhandled exception at 0x00007FFE1B196B76 (nppdf32.dll) in airapp.exe: 0xC0000005: Access violation reading location 0x000000007154DD10

This library is loaded by webkit.

**Reply by concord232** [reply]
Were you able to find any solution? Thanks!

**Reply by ajwfrost** [reply]
There were two parts to the problem that we found. The first was a crash that we did fix, but then it resulted in a second crash when trying to communicate between the AIR app (more specifically, the JavaScript hosted within the WebKit library) and the NPP plug-in from Adobe Acrobat. It seems Acrobat have dropped support for the interactivity, there are just a few limited things that you can do via this API (basically limited to displaying the PDF). If you wanted more capabilities then it would need to be using the full subscription-based Acrobat DC rather than the free Reader version, and it would need to be done via a different API.

So depending on what you're trying to do, it may be better to use StageWebView and the Edge WebView2 component to display PDF files within the Edge native PDF viewer, rather than trying to use Acrobat Reader.

thanks

**Reply by pcollardez** [reply]
Indeed, StageWebView with Edge WebView2 is the only option.
Some remarks about this :

- the WebView always overlap the AIR window, some specific wiring is required to detect AIR events (resize, menus, popus, dividers...) and switch to a snapshot of the WebView window or to the WebView window (by simply detaching/attaching the WebView to the Stage)
- on Windows, the snapshot function is always async (drawViewPortAsync) and only available since player version 44 (see your compiler options)
- the StageWebView must not be detached from the Stage before the event WEBVIEW_DRAW_COMPLETE is handled (it seems to create a screen capture, the WebView window must be displayed to work properly; if the StageWebView is detached too early, the event never fires)
